File: impl/go/point_lookup_test.go
package jed

// Primary-key predicate pushdown over a MULTI-LEAF B-tree (spec/design/cost.md §3 "bounded scan").
// The conformance corpus (spec/conformance/suites/query/point_lookup.test) pins the cost contract on
// single-leaf tables; this exercises what it cannot — a tree tall/wide enough that page_read actually
// drops below the full node count, and a range scan that spans leaf boundaries.

import (
	"fmt"
	"strings"
	"testing"
)

// bigTable builds a table of n rows (id i32 PRIMARY KEY, v i32; v == id) large enough to span
// several B-tree leaves at the default page size.
func bigTable(t *testing.T, n int) *engine {
	t.Helper()
	var b strings.Builder
	b.WriteString("INSERT INTO t VALUES ")
	for i := 1; i <= n; i++ {
		if i > 1 {
			b.WriteByte(',')
		}
		fmt.Fprintf(&b, "(%d,%d)", i, i)
	}
	return dbWith(t, "CREATE TABLE t (id i32 PRIMARY KEY, v i32)", b.String())
}

func TestPointLookupMultiLeaf(t *testing.T) {
	const n = 1000
	db := bigTable(t, n)

	store := db.readSnap().store("t")
	nodeCount := store.NodeCount()
	if nodeCount <= 1 {
		t.Fatalf("test needs a multi-leaf tree; got nodeCount=%d (raise n)", nodeCount)
	}

	// Storage primitive: a point bound visits strictly fewer nodes than the whole tree (the page_read
	// win), and returns exactly the one matching row.
	key := encodeInt(scalarInt32, 500)
	pb := keyBound{lo: key, loInc: true, hi: key, hiInc: true}
	if got := store.OverlapNodeCount(pb); got >= nodeCount {
		t.Errorf("point-lookup overlap node count %d should be < full node count %d", got, nodeCount)
	}
	rows, err := store.RangeRows(pb)
	if err != nil {
		t.Fatal(err)
	}
	if len(rows) != 1 || rows[0][0].Int != 500 {
		t.Errorf("RangeRows point bound got %v, want one row id=500", rows)
	}

	// End-to-end point lookup: exactly one row, and the cost is sublinear in n (it did not full-scan).
	out, err := execute(db, "SELECT v FROM t WHERE id = 500")
	if err != nil {
		t.Fatal(err)
	}
	if len(out.Rows) != 1 || out.Rows[0][0].Int != 500 {
		t.Errorf("point lookup got %v, want [500]", out.Rows)
	}
	full, err := execute(db, "SELECT v FROM t")
	if err != nil {
		t.Fatal(err)
	}
	if out.Cost >= full.Cost {
		t.Errorf("point-lookup cost %d should be far below full-scan cost %d", out.Cost, full.Cost)
	}
	if out.Cost > 50 {
		t.Errorf("point-lookup cost %d should be small (≈ tree height + a few), not ~%d", out.Cost, n)
	}

	// A point-lookup MISS still visits the leaf it would live in (page_read > 0) but reads no row.
	miss, err := execute(db, "SELECT v FROM t WHERE id = 99999")
	if err != nil {
		t.Fatal(err)
	}
	if len(miss.Rows) != 0 || miss.Cost == 0 || miss.Cost > 50 {
		t.Errorf("point-lookup miss got rows=%d cost=%d, want 0 rows and a small non-zero cost", len(miss.Rows), miss.Cost)
	}
}

func TestRangeScanCrossesLeafBoundaries(t *testing.T) {
	const n = 1000
	db := bigTable(t, n)

	// A range that spans many leaves returns the contiguous, correct rows in key order — the property
	// the single-leaf corpus cannot exercise (the leaf-crossing in-order traversal).
	got := queryIDs(t, db, "SELECT id FROM t WHERE id >= 300 AND id <= 700 ORDER BY id")
	if len(got) != 401 {
		t.Fatalf("range scan got %d rows, want 401", len(got))
	}
	for i, id := range got {
		if id != int64(300+i) {
			t.Fatalf("range row %d = %d, want %d", i, id, 300+i)
		}
	}

	// Open-ended range to the end of the key space.
	tail := queryIDs(t, db, "SELECT id FROM t WHERE id > 996 ORDER BY id")
	if len(tail) != 4 || tail[0] != 997 || tail[3] != 1000 {
		t.Errorf("open range got %v, want [997..1000]", tail)
	}

	// Empty (contradictory) bound: zero rows, zero cost — proved without scanning.
	empty, err := execute(db, "SELECT id FROM t WHERE id > 700 AND id < 300")
	if err != nil {
		t.Fatal(err)
	}
	if len(empty.Rows) != 0 || empty.Cost != 0 {
		t.Errorf("empty bound got rows=%d cost=%d, want 0 rows and cost 0", len(empty.Rows), empty.Cost)
	}
}

func TestLimitShortCircuitMultiLeaf(t *testing.T) {
	const n = 1000
	db := bigTable(t, n) // id 1..1000, v == id

	// LIMIT without ORDER BY stops the scan early: it returns `limit` rows at sublinear cost (it did
	// NOT read all 1000). The rows are the primary-key-order prefix (our cores' deterministic choice).
	out, err := execute(db, "SELECT v FROM t LIMIT 5")
	if err != nil {
		t.Fatal(err)
	}
	if len(out.Rows) != 5 {
		t.Fatalf("LIMIT 5 got %d rows, want 5", len(out.Rows))
	}
	full, err := execute(db, "SELECT v FROM t")
	if err != nil {
		t.Fatal(err)
	}
	if out.Cost >= full.Cost {
		t.Errorf("LIMIT cost %d should be far below the full-scan cost %d", out.Cost, full.Cost)
	}
	if out.Cost > 20 {
		t.Errorf("LIMIT 5 cost %d should be sublinear (≈ limit + node count), not ~%d", out.Cost, n)
	}
	got := queryIDs(t, db, "SELECT v FROM t LIMIT 5")
	for i, v := range got {
		if v != int64(i+1) {
			t.Fatalf("LIMIT 5 row %d = %d, want %d", i, v, i+1)
		}
	}

	// OFFSET skips then takes, still sublinear.
	o := queryIDs(t, db, "SELECT v FROM t LIMIT 3 OFFSET 10")
	if len(o) != 3 || o[0] != 11 || o[2] != 13 {
		t.Errorf("LIMIT 3 OFFSET 10 got %v, want [11 12 13]", o)
	}

	// Trap windowing: streaming projects ONLY the windowed rows (like the eager path), so a later
	// row that would trap is never reached under a LIMIT that excludes it.
	dz := dbWith(t,
		"CREATE TABLE z (id i32 PRIMARY KEY, c i32)",
		"INSERT INTO z VALUES (1, 5), (2, 0), (3, 5)")
	if rows := query(t, dz, "SELECT 100 / c FROM z LIMIT 1"); len(rows) != 1 || rows[0][0].Int != 20 {
		t.Errorf("LIMIT 1 should produce only the safe first row (100/5=20), got %v", rows)
	}
	if _, err := execute(dz, "SELECT 100 / c FROM z LIMIT 2"); err == nil {
		t.Errorf("LIMIT 2 reaches the c=0 row and must trap (division by zero)")
	}
}

func TestMutationPushdownMultiLeaf(t *testing.T) {
	const n = 1000
	db := bigTable(t, n)

	// DELETE by primary key seeks one row at sublinear cost and leaves the rest intact.
	d, err := execute(db, "DELETE FROM t WHERE id = 500")
	if err != nil {
		t.Fatal(err)
	}
	if d.Cost > 50 {
		t.Errorf("DELETE point-lookup cost %d should be small (sublinear in %d)", d.Cost, n)
	}
	if got := queryIDs(t, db, "SELECT id FROM t WHERE id = 500"); len(got) != 0 {
		t.Errorf("row 500 should be deleted, got %v", got)
	}
	if got := queryIDs(t, db, "SELECT id FROM t WHERE id = 501"); len(got) != 1 {
		t.Errorf("neighbouring rows must survive the point delete")
	}

	// UPDATE by primary key likewise.
	u, err := execute(db, "UPDATE t SET v = -1 WHERE id = 700")
	if err != nil {
		t.Fatal(err)
	}
	if u.Cost > 50 {
		t.Errorf("UPDATE point-lookup cost %d should be small", u.Cost)
	}
	if got := query(t, db, "SELECT v FROM t WHERE id = 700"); got[0][0].Int != -1 {
		t.Errorf("row 700 should be updated to -1, got %v", got[0][0].Int)
	}
}

File: impl/ts/tests/point_lookup.test.ts
// Primary-key predicate pushdown over a MULTI-LEAF B-tree (spec/design/cost.md §3 "bounded scan").
// The conformance corpus (spec/conformance/suites/query/point_lookup.test) pins the cost contract on
// single-leaf tables; this exercises what it cannot — the bounded-scan primitive where page_read drops
// below node_count, and a range scan that spans leaf boundaries.

import assert from "node:assert/strict";
import { test } from "node:test";
import { Engine, execute, intValue } from "../src/tooling.ts";
import { PMap, unboundedBound } from "../src/pmap.ts";
import type { KeyBound } from "../src/pmap.ts";
import type { Row } from "../src/storage.ts";
import type { Value } from "../src/value.ts";

// --- direct storage-primitive check (the page_read drop) ---

const CAP = 240;
const W = 15;

function key(n: number): Uint8Array {
  const b = new Uint8Array(8);
  new DataView(b.buffer).setBigUint64(0, BigInt(n));
  return b;
}

function pnodeRow(n: number): Row {
  return [intValue(BigInt(n))];
}

test("bounded range + overlap over a multi-leaf tree", () => {
  const m = new PMap();
  for (let n = 0; n < 200; n++) m.insert(key(n), pnodeRow(n), W, CAP, null);
  assert.ok(m.nodeCount() > 1, "test needs a multi-leaf tree");

  // A point bound visits strictly fewer nodes than the whole tree (the page_read win).
  const pb: KeyBound = { lo: key(100), loInc: true, hi: key(100), hiInc: true };
  assert.ok(m.overlapNodeCount(pb) < m.nodeCount());
  assert.strictEqual(m.rangeEntries(pb, null).vals.length, 1);

  // An inclusive range spanning many leaves returns exactly those entries (50..=150 = 101).
  const rb: KeyBound = { lo: key(50), loInc: true, hi: key(150), hiInc: true };
  assert.strictEqual(m.rangeEntries(rb, null).vals.length, 101);

  // Exclusive endpoints drop both boundary keys (51..=149 = 99).
  const ex: KeyBound = { lo: key(50), loInc: false, hi: key(150), hiInc: false };
  assert.strictEqual(m.rangeEntries(ex, null).vals.length, 99);

  // Half-open (>= 195) reaches the end of the key space (195..=199 = 5).
  const hiOpen: KeyBound = { lo: key(195), loInc: true, hi: null, hiInc: false };
  assert.strictEqual(m.rangeEntries(hiOpen, null).vals.length, 5);

  // The unbounded bound reproduces the full scan exactly.
  const unb = unboundedBound();
  assert.strictEqual(m.overlapNodeCount(unb), m.nodeCount());
  assert.strictEqual(m.rangeEntries(unb, null).vals.length, 200);
});

// --- end-to-end (public API): correctness across leaves + sublinear cost ---

function bigTable(n: number): Engine {
  const db = new Engine();
  execute(db, "CREATE TABLE t (id i32 PRIMARY KEY, v i32)");
  const parts: string[] = [];
  for (let i = 1; i <= n; i++) parts.push(`(${i},${i})`);
  execute(db, "INSERT INTO t VALUES " + parts.join(","));
  return db;
}

function cost(db: Engine, sql: string): bigint {
  return execute(db, sql).cost;
}

function intOf(v: Value): number {
  if (v.kind !== "int") throw new Error("expected int, got " + v.kind);
  return Number(v.int);
}

function ids(db: Engine, sql: string): number[] {
  const o = execute(db, sql);
  if (o.kind !== "query") throw new Error("expected a query result");
  return o.rows.map((r) => intOf(r[0]));
}

test("point lookup is sublinear", () => {
  const db = bigTable(1000);
  assert.deepStrictEqual(ids(db, "SELECT v FROM t WHERE id = 500"), [500]);
  const point = cost(db, "SELECT v FROM t WHERE id = 500");
  const full = cost(db, "SELECT v FROM t");
  assert.ok(point < full, `point cost ${point} should be far below full-scan ${full}`);
  assert.ok(point <= 50n, `point cost ${point} should be small (≈ height + a few), not ~1000`);
  assert.deepStrictEqual(ids(db, "SELECT v FROM t WHERE id = 99999"), []);
  const miss = cost(db, "SELECT v FROM t WHERE id = 99999");
  assert.ok(miss > 0n && miss <= 50n, `miss cost ${miss} should be small and non-zero`);
});

test("range scan crosses leaf boundaries", () => {
  const db = bigTable(1000);
  const got = ids(db, "SELECT id FROM t WHERE id >= 300 AND id <= 700 ORDER BY id");
  assert.strictEqual(got.length, 401);
  for (let i = 0; i < got.length; i++) assert.strictEqual(got[i], 300 + i);
  assert.deepStrictEqual(
    ids(db, "SELECT id FROM t WHERE id > 996 ORDER BY id"),
    [997, 998, 999, 1000],
  );
  // Contradictory bound: empty, cost 0.
  assert.deepStrictEqual(ids(db, "SELECT id FROM t WHERE id > 700 AND id < 300"), []);
  assert.strictEqual(cost(db, "SELECT id FROM t WHERE id > 700 AND id < 300"), 0n);
});

test("LIMIT short-circuit is sublinear", () => {
  const db = bigTable(1000); // id 1..1000, v == id
  // LIMIT without ORDER BY stops the scan early: `limit` rows at sublinear cost, the PK-order prefix.
  assert.deepStrictEqual(ids(db, "SELECT v FROM t LIMIT 5"), [1, 2, 3, 4, 5]);
  const point = cost(db, "SELECT v FROM t LIMIT 5");
  const full = cost(db, "SELECT v FROM t");
  assert.ok(point < full, `LIMIT cost ${point} should be far below full-scan ${full}`);
  assert.ok(
    point <= 20n,
    `LIMIT 5 cost ${point} should be sublinear (≈ limit + node count), not ~1000`,
  );
  assert.deepStrictEqual(ids(db, "SELECT v FROM t LIMIT 3 OFFSET 10"), [11, 12, 13]);

  // Trap windowing: streaming projects ONLY the windowed rows, so a later trapping row is never
  // reached under a LIMIT that excludes it.
  const dz = new Engine();
  execute(dz, "CREATE TABLE z (id i32 PRIMARY KEY, c i32)");
  execute(dz, "INSERT INTO z VALUES (1, 5), (2, 0), (3, 5)");
  assert.deepStrictEqual(ids(dz, "SELECT 100 / c FROM z LIMIT 1"), [20]);
  assert.throws(() => execute(dz, "SELECT 100 / c FROM z LIMIT 2"));
});

test("mutation pushdown is sublinear", () => {
  const db = bigTable(1000);
  const d = cost(db, "DELETE FROM t WHERE id = 500");
  assert.ok(d <= 50n, `DELETE point-lookup cost ${d} should be sublinear`);
  assert.deepStrictEqual(ids(db, "SELECT id FROM t WHERE id = 500"), []);
  assert.deepStrictEqual(ids(db, "SELECT id FROM t WHERE id = 501"), [501]);

  const u = cost(db, "UPDATE t SET v = -1 WHERE id = 700");
  assert.ok(u <= 50n, `UPDATE point-lookup cost ${u} should be sublinear`);
  assert.deepStrictEqual(ids(db, "SELECT v FROM t WHERE id = 700"), [-1]);
});
